### ISSUE TITLE
Proposal: Factory functions for string and bytes types

### ISSUE THREAD

**Issue by e-kayrakli** [chapel-lang/chapel#13818]
This issue is to propose adding factory functions to create strings and bytes.
The current initializers are arguably confusing. The confusion is mainly about
the ownership transfer of the underlying byte buffers.

Current Initializers
--------------------

Below are current `string` initializers. `bytes` type also have identical
initializers.

- `proc init(s: string, isowned: bool = true)`

  If `isowned` is set, `this` copies `s`'s buffer and "owns" it. That means,
  `this` is responsible of its deallocation.

- `proc init(cs: c_string, length:int=cs.length, isowned:bool=true, needToCopy: bool=true)`
- `proc init(buff: bufferType, length:int, size:int, isowned:bool=true, needToCopy:bool=true)`

  I tried to create a truth-table to list the behavior with different values of
  flags but I gave up. Current implementation uses a `reinitString` method for
  both assignment and initialization, which doesn't seem like the right approach
  to me. That may be the reason why we have this interface here.

  The doc for this initializer says:

  ```
      Initialize a new string from `buff` ( `c_ptr(uint(8))` ). `size` indicates
      the total size of the buffer available, while `len` indicates the current
      length of the string in the buffer (the common case would be `size-1` for
      a C-style string). If `isowned` is set to true, the backing buffer will be
      freed when the new record is destroyed. If `needToCopy` is set to true,
      the `c_string` will be copied into the record, otherwise it will be used
      directly. It is the responsibility of the user to ensure that the
      underlying buffer is not freed if the `c_string` is not copied in.
  ```
  Followed by a comment: 
  ```
  This initializer can cause a leak if isowned = false and needToCopy = true
  ```

What do we need?
-----------------

We need to be able to create strings (and bytes) with the following ownership
transfers:

- New object "borrows" the pointer.
  No need to create a copy or deallocate.

- New object takes ownership of the pointer.
  No need to copy. Deallocate during deinit.

- New object copies the buffer.
  Create a copy and deallocate.

What is proposed?
-----------------

I propose the following factory functions that return string/bytes objects.

- createStringWithBorrowedBuffer (no copy, no deallocate)
- createStringWithOwnedBuffer (no copy, yes deallocate)
- createStringWithNewBuffer (yes copy, yes deallocate)
- createBytesWithBorrowedBuffer
- createBytesWithOwnedBuffer
- createBytesWithNewBuffer

I am not too crazy about these names as they are long. But they should be descriptive.

What will happen to existing initializers?
------------------------------------------

We will have
- `proc init(other: string)`
- `proc init=(other:string)`
That'll create a copy of `s`'s buffer

**Comment by mppf** [comment]
> createStringWithBorrowedBuffer (no copy, no deallocate)
> createStringWithOwnedBuffer (no copy, yes deallocate)
> createStringWithNewBuffer (yes copy, yes deallocate)

How about

`createStringBorrowingBuffer`
`createStringOwningBuffer`
`createStringCopyingBuffer`

?

**Comment by bradcray** [comment]
Which of these features are intended to be user-facing, and which are intended for internal- / developer- (possibly power-user-) only purposes?

What's the advantage of having `string.init(s: string)` rather than (or in addition to) `string.init=(s: string)`?

**Comment by e-kayrakli** [comment]
>Which of these features are intended to be user-facing, and which are intended for internal- / developer- (possibly power-user-) only purposes?

I intend these to be user-facing. However, one might also say trying to create a Chapel string from a C array is a power-user operation? Currently if you want to do that, you'll have to use these initializers. Which, I argue, does not have a great interface.

> What's the advantage of having string.init(s: string) rather than (or in addition to) string.init=(s: string)?

If we don't have the former, then we won't be able to do `new string("foo")`? In other words, don't we need both to be able to do the following?
```chapel
var s1 = new string("foo");
var s1 = "bar";
```

**Comment by bradcray** [comment]
I'd definitely consider creating a Chapel string from a C array to be a power user feature.

I'm wondering whether we need/want to be able to do `new string("foo")` as opposed to just `var s1 = "foo";` if we're viewing `string` as a fairly primitive type for the (typical) user.  By analogy, we don't support `new int(42)` for example (that said, we have talked about potentially wanting to support that for ints for orthogonality...  but it would require converting their implementation to use records or something else special).

**Comment by e-kayrakli** [comment]
> I'd definitely consider creating a Chapel string from a C array to be a power user feature.

I guess it is still safe to say that these functions are intended to be user-facing. However, I'd be surprised to see a lot of user code relying on these.

One internal usage (that I am aware of) for the current initializers is in the IO module.

> I'm wondering whether we need/want to be able to do new string("foo") as opposed to just var s1 = "foo"; if we're viewing string as a fairly primitive type for the (typical) user. By analogy, we don't support new int(42) for example (that said, we have talked about potentially wanting to support that for ints for orthogonality... but it would require converting their implementation to use records or something else special).

I think this is a separate discussion and whatever route we choose shouldn't affect the decisions here. But I may be missing something.

**Comment by BryantLam** [comment]
> I tried to create a truth-table to list the behavior with different values of flags but I gave up.

I welcome this proposal because I share this sentiment. The `string.init(c_string)` initializer is difficult to reason about for library developers doing interoperability to C libraries, which I imagine is a common task, not solely one done by power users. Maybe those library developers are closer to power users than typical programmers.

I also dislike setting bool values on an initializer that changes how the type behaves, since that feels like it should just be its own type.* This proposal gets around that feeling by providing factory functions that do exactly the behavior that I want, not the behavior that I have to remember to set correctly on an initializer with default argument values.

\* In a what-if digression, the managed types seem like they could have been aptly applied here in an alternative history.
```chapel
var sbrw: borrowed string = c_string;
var sown = new owned string(c_string);
var snew: string = c_string;
```

**Comment by bradcray** [comment]
> I think this is a separate discussion and whatever route we choose shouldn't affect the decisions here. But I may be missing something.

I got to it from this comment:

> We will probably only have init(s: string) which takes ownership of the buffer of s.

And wondering whether we could simply avoid supporting `init(s: string)` as well.

I worry a little about whether "takes ownership of the buffer of s" would be confusing default behavior to a typical user.  That is, my intuition for the following code:

```chapel
var s1 = "hi";
var s2 = new string(s1);  // This suggests to me: give me a new string initialized to a copy of s1, not to take ownership of s1
```

is that s2 would be a new string that was initialized to a copy of s1 (so, equivalent to `var s2 = s1;` or `var s2 = "hi";`).  Is there some way in which I could get into trouble if I thought it was doing a deep copy of the string (similar to `init=`, I'm guessing?) but it was transferring ownership instead?

**Comment by mppf** [comment]
> And wondering whether we could simply avoid supporting init(s: string) as well.

This would be fine with me.

> I worry a little about whether "takes ownership of the buffer of s" would be confusing default behavior to a typical user. That is, my intuition for the following code:

I think @e-kayrakli mischaracterized the current implementation. AFAIK `var s2 = new string(s1)` will create a copy of the string's buffer.

**Comment by BryantLam** [comment]
> > createStringWithBorrowedBuffer (no copy, no deallocate)
> > createStringWithOwnedBuffer (no copy, yes deallocate)
> > createStringWithNewBuffer (yes copy, yes deallocate)
> 
> How about
> 
> `createStringBorrowingBuffer`
> `createStringOwningBuffer`
> `createStringCopyingBuffer`
> 
> ?

I prefer the original proposal's style of naming. `borrowed`, `owned`, and `new` are already keywords I need to be familiar with to work with classes. I shouldn't need to remember that string factory functions are named `Borrowing`, `Owning`, and `Copying` also. Maybe the original naming is too wordy, but it's fine to me; the calls will usually be the only expression on a statement, unless someone does something crazy like:
`funcTakesConstRefStrings(createStringWithBorrowedBuffer(cs1), createStringWithBorrowedBuffer(cs2));`
which isn't totally out of question with that specific function.

Taking my alternative history a bit farther, another proposal could be like `createOwnedString` .. `createNewString`*. I suggest it only semi-seriously since it also has a 'quirk' in that factory functions are usually `createType` though this "type" is technically an `owned string`.

\*  (or just `createString` but why not just use the default initializer for this case).

**Comment by e-kayrakli** [comment]
> I think @e-kayrakli mischaracterized the current implementation. AFAIK var s2 = new string(s1) will create a copy of the string's buffer.

You are right. I got confused.

**Comment by vasslitvinov** [comment]
Indeed, if we made String a class, then we'd get all the the ownership machinery for free? As Bryant suggested. The downside I see is the need in some compiler magic to avoid double-indirection to get at the underlying buffer.

**Comment by mppf** [comment]
> Indeed, if we made String a class, then we'd get all the the ownership machinery for free? As Bryant suggested. The downside I see is the need in some compiler magic to avoid double-indirection to get at the underlying buffer.

There are other problems than that. The main problem I see is that records can define `init=` and `=` overloads to support these operations. But you can't redefine what `owned` or `shared` do in this regard.

The main impact of that for string-as-class would be that we'd need another "ownership" type for "copy the pointed-to-data on assignment".

Anyway I'm not sure that making `owned string` and `shared string` really helps make user code clearer. I have sometimes wished for `borrowed string` (but that could be another type, like `borrowedString`) to represent things like string slices without requiring the implementation to always copy.

**Comment by BryantLam** [comment]
It's a nice idea though I definitely think there's too many trade-offs to consider it seriously. Hence why I suggested it semi-seriously. ;)

**Comment by dlongnecke-cray** [comment]
```chapel
enum StringInitStyle {
    New,
    Owned,
    Borrowed
}

proc init(buffer: c_string, style: StringInitStyle=New) { ... }
```

Would be preferable to me over an overly verbose factory function, if given the choice.

Though I rather like Vass's idea of just making `string` a class and getting all of this ownership machinery for free...

**Comment by vasslitvinov** [comment]
Ah right, we want to present `string` as a primitive type like `int`, so the user shouldn't be concerned about ownership issues. Except when creating a new string in a non-standard way, as a power user might want to.

**Comment by mppf** [comment]
> proc init(buffer: c_string, style: StringInitStyle=New) { ... }

I definitely prefer the separate functions over this strategy.

**Comment by BryantLam** [comment]
The enum version is mutually exclusive. I wouldn't mind if the enum version replaced the current initializers' use of bools. Reading a function call with `true` or `false` in it is maddening without IDE support to tell me what the formals are.

I still want the factory functions though.
